CodeU_Assignement_5/src/codeu_assignement_5/KemUnknownLanguage.java

```diff
@@ -0,0 +1,145 @@
+        /* Get the current order by prefix size, starting from 0 */
+        List<Character> dict = getBasicAlphaBeit(alienWords, 0); /* Basic Alphabeit */ 
+        /* A,R,C from the example! */
+
+        /* Add from prefix while there are letters after first letter of prefix */
+        while (!addToDictByPrefix(alienWords, dict)){} 
+
+        /* Add all the chars we cannot decide in the end*/
+        for (String w : alienWords) {
+            for (Character c: w.toCharArray()) {
+                if (!dict.contains(c)) {
+                    dict.add(c);
+                }
+            }
+        }
+        
+        return dict;
+    }
+    
+    
+    /*
+     * Inputs: All words with the same prefix (all the first prefixSize chars are the same!)
+     * Output: Per common prefix, gets the order of the chars
+     *
+     * Note: Assume the input is valid here!
+    */
+    private List<Character> getBasicAlphaBeit(String[] words, int prefixSize) {
+        List<Character> ret = new ArrayList(); 
+        if (words == null) {
+            return ret;
+        }
+        
+        for (String w : words) {
+            if ((w.length() > prefixSize) && (!ret.contains(w.charAt(prefixSize)))) {
+                ret.add(w.charAt(prefixSize));
+            }
+        }   
+        
+        return ret;
+    }
+    
+    /*
+     * Gives the prefix of a char that isn't in the dictionary
+    */
+    private int getSizeOfPrefix(String w, List<Character> dict) {
+        for (int i=0; i < w.length(); i++) {
+            if (!dict.contains(w.charAt(i))) {
+                return i;
+            }
+        }
+        
+        return w.length();
+    }
+    
+    /*
+     * Input: the original set of words, current word we are working on, its prefix size
+     * Output: the set of the words with this prefix
+    */
+    private String[] getWordsWithSamePrefix(String[] alienWords, int curr, int prefixSize) {
+        ArrayList<String> tempDict = new ArrayList();
+        tempDict.add(alienWords[curr]);
+        
+        String prefix = alienWords[curr].substring(0, prefixSize);
+        
+        /* Forward search */
+        for (int i=curr+1; i<alienWords.length; i++) {
+            if (!alienWords[i].startsWith(prefix)) {
+                break;
+            }
+            
+            tempDict.add(alienWords[i]);
+        }
+        
+        /* Backward search */
+        for (int i=curr-1; i >= 0; i--) {
+            if (!alienWords[i].startsWith(prefix)) {
+                break;
+            }
+            
+            tempDict.add(0,alienWords[i]); 
+        }
+        
+        return tempDict.toArray(new String[tempDict.size()]);
+    }
+    
+    /*
+     * Get two dictionary with common latters and merge them into the first dict,
+     * if none, cannot know, add all these cases later on
+    */
+    private void mergeDict(List<Character> dict1, List<Character> dict2) {
```

Comment:
These are potential *alphabets*, not *dictionaries*. I guess the comment could be:

```
/* Given two sub-alphabets, generates an (arbitrary) merged alphabet which is consistent with both of them. */
```

I think this method is the core of my problem with the algorithm. Given two sub-alphabets (e.g. ABD, ACD) there may be multiple potential alphabets that satisfy both (e.g. ABCD, ACBD). This algorithm will choose one, and then a later pass may reveal whether the ordering is actually BC or CB.

In other words, this method takes two orderings and produces a *potential* ordering, but it might be the wrong one based on information revealed later.

Your code may do subtle things to make this work, but I see no comments or structures in the code to suggest you've addressed this issue.